cobc/parser.y

```diff
@@ -8087,6 +8135,17 @@ usage:
   }
 ;
 
+_to_program_type:
+  /* empty */		{ $$ = NULL; }
+| _to PROGRAM_NAME	{ $$ = $2;   }
+;
+
+_to_type_name:
+  /* empty */ %prec SHIFT_PREFER { $$ = NULL; }
+| TYPEDEF_NAME                   { $$ = $1;   }
+| TO TYPEDEF_NAME                { $$ = $2;   }
```

Comment:
> If there's a shift-conflict with that, please add a TODO entry for later inspection.

I inspected that already, hence the proposed changes.

There are actually two conflicts, a reduce/reduce, and a shift/reduce.

The reduce/reduce conflict is easy to understand considering these three rules:
```
usage:
...
| POINTER _to_type_name { ... }
...
| TYPEDEF_NAME { ... }
...

_to_type_name:
  /* empty */ { ... }
| _to TYPEDEF_NAME { ... }

_to:
  /* empty */
| TO ;
```

So when you have `POINTER SOME-TYPEDEF-NAME`, after shifting `POINTER` and when looking ahead `SOME-TYPEDEF-NAME`, we can't decide if we should reduce `_to_type_name -> epsilon` or `_to -> epsilon`. The only way I know to resolve these kind of conflicts is by inlining rules (in OCaml's parser generator (Menhir) we even have an `%inline` annotation we can add to rules to have them automatically inlined :p).

But then, even if we solve this reduce/reduce conflict, `POINTER SOME-TYPEDEF-NAME` is still ambiguous (shift/reduce conflict): is `SOME-TYPEDEF-NAME` the argument of `POINTER` (which makes sense), or is it a new `USAGE` clause (which does not make sense but is actually allowed by the grammar, since we just have a list of clauses, they can be repeated). Hence the need to give a precedence to `TYPEDEF_NAME` and the `%prec SHIFT_PREFER` annotation on the epsilon production in `_to_type_name` - to favor the first case.

Reply:
What about

~~~
usage:
...
| POINTER _to TYPEDEF_NAME { ... }
| POINTER  { ... }
...
| TYPEDEF_NAME { ... }
```
?

Reply:
That still makes a reduce/reduce conflict (tested): after shifting `POINTER` and when looking ahead at `TYPEDEF-NAME`, we can't decide if we should reduce `usage -> POINTER` or `_to -> epsilon`.

Reply:
By "here", do you mean in the `usage` rule ?
(if so, since it's a reduce/reduce conflict, precedence will not help)

Reply:
as noted: as long as you have a solution here we go with it now - but as this part isn't covered by the testsuite (should be in syn_definition), I suggest to add the []PROGRAM-/FUNCTION-]POINTER TO ... TYPE_NAME/FUNCTION_NAME/PROGRAM_NAME with and without TO - this way we ensure that it works now and later.

... you may commit this to 3.x and merge or merge if that exists in 3.x already.

Reply:
Note the 3.x branch does not have a conflict here (since it does not allow specifying a TYPEDEF_NAME as a USAGE).

> ... you may commit this to 3.x and merge or merge if that exists in 3.x already.
Are you talking about the additional POINTER tests ?

Reply:
I confirm that the test I added to #178 has the exact same result here.